src/relay/ir/doc.h

```diff
@@ -0,0 +1,436 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file doc.h
+ * \brief A pretty printer DSL for constructing (Doc) and formatting (RDoc) documents.
+ *        It is based heavily on Philip Wadler's "A prettier printer."
+ *        See https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf
+ *        for more details.
+ *
+ * Since the original paper uses call by value for efficiency, everything doc function is maximally lazy.
+ * You can probably yank speed by doing strict analysis and removing some Lazy (if this is bottleneck).
+ */
+#ifndef TVM_RELAY_IR_DOC_H_
+#define TVM_RELAY_IR_DOC_H_
+
+#include <map>
+#include <utility>
+#include <string>
+#include <functional>
+#include <vector>
+#include <memory>
+#include <ostream>
+#include "error.h"
+
+namespace tvm {
+namespace relay {
+
+struct DocError : public Error {
+  explicit DocError(const std::string& msg) : Error(msg) { }
+};
+
+template<typename T>
+struct LazyNode {
+  mutable std::function<T()> thunk;
+  explicit LazyNode(const std::function<T()>& thunk) : thunk(thunk) { }
+};
+
+//! \brief denote a value that will be computed (at most once) on need.
+template<typename T>
+struct Lazy {
+  std::shared_ptr<LazyNode<T>> ln;
+  explicit Lazy(const std::function<T()>& thunk) : ln(std::make_shared<LazyNode<T>>(thunk)) { }
+  explicit Lazy(const T& value) : Lazy([=]() { return value; }) { }
+  explicit Lazy(const Lazy<Lazy<T>>& thunk) : Lazy([=]() { return thunk.get().get(); }) { }
+  T get() const {
+    T res = ln->thunk();
+    ln->thunk = [=]() { return res; };
+    return res;
+  }
+  template<typename R>
+  Lazy<R> map(const std::function<R(const T&)>& func) const {
+    Lazy<T> self(*this);
+    return Lazy<R>([=]() -> R { return func(self.get()); });
+  }
+};
+
+struct NilNode;
+struct AppNode;
+struct NestNode;
+struct TextNode;
+struct LineNode;
+struct ChoiceNode;
+
+/*! \brief The inner representation of Doc.
+ * a doc represent structured text,
+ * and can be rendered onto screen while keeping the structure.
+ */
+struct DocNode {
+  std::shared_ptr<NilNode> nil;
+  std::shared_ptr<AppNode> app;
+  std::shared_ptr<NestNode> nest;
+  std::shared_ptr<TextNode> text;
+  std::shared_ptr<LineNode> line;
+  std::shared_ptr<ChoiceNode> choice;
+  DocNode(std::shared_ptr<NilNode> nil,
+           std::shared_ptr<AppNode> app,
+           std::shared_ptr<NestNode> nest,
+           std::shared_ptr<TextNode> text,
+           std::shared_ptr<LineNode> line,
+           std::shared_ptr<ChoiceNode> choice) :
+    nil(nil),
+    app(app),
+    nest(nest),
+    text(text),
+    line(line),
+    choice(choice) { }
+};
+
+struct Doc {
+  Lazy<DocNode> doc;
+  explicit Doc(const DocNode& ed) : doc(ed) { }
+  explicit Doc(const Lazy<Doc>& ldoc) :
+    doc(ldoc.map<Lazy<DocNode>>([](const Doc& d){ return d.doc; })) { }
+
+  Doc operator+(const Doc&) const;
+
+  template<typename T>
+  Lazy<T> match(
```

Comment:
document the parameter of match function.

match->Match (google C style)

Reply:
done